Tryouts/Messaging-JS/src/client/MessageRouterClient.ts

```diff
@@ -0,0 +1,447 @@
+                requestId: message.requestId,
+                payload: typeof result === "string" ? result : undefined
+            });
+
+        } catch (error) {
+
+            await this.sendMessage<messages.InvokeResponse>({
+                type: "InvokeResponse",
+                requestId: message.requestId,
+                error: createProtocolError(error)
+            });
+        }
+    }
+
+    private handleConnectResponse(message: messages.ConnectResponse): void {
+        if (message.error) {
+            this._state = ClientState.Closed;
+            this.connected.reject(new MessageRouterError(message.error));
+        }
+        else {
+            this._state = ClientState.Connected;
+            this.clientId = message.clientId;
+            this.connected.resolve();
+        }
+    }
+
+    private checkState() {
+        if (this._state == ClientState.Closed || this._state == ClientState.Closing) {
+            throw ThrowHelper.connectionClosed();
+        }
+    }
+
+    private async handleError(err: any): Promise<void> {
+        this.failPendingRequests(err);
+        this.failSubscribers(err);
+        
+        if (this._state == ClientState.Connecting) {
+            this.connected.reject(err);
+            return;
+        }
+
+        this._state = ClientState.Closed;
+        await this.connection?.close();
+    }
+
+    private handleClose(): Promise<void> {
+        switch (this._state) {
+            case ClientState.Closing:
+            case ClientState.Closed:
+                return Promise.resolve();
+        }
+
+        return this.handleError(ThrowHelper.connectionClosed());
+    }
+
+    private async unsubscribe(topicName: string): Promise<void> {
+        let topic = this.topics[topicName];
+
+        if (!topic)
+            return;
+
+        await this.sendMessage<messages.UnsubscribeMessage>(
+            {
+                type: "Unsubscribe",
+                topic: topicName
+            }
+        );
+    }
+
+    private getRequestId(): string {
+        return '' + (++this.lastRequestId);
+    }
+}
+
+class Topic {
+
+    constructor(onUnsubscribe: () => void) {
+        this.onUnsubscribe = onUnsubscribe;
+    }
+
+    subscribe(subscriber: TopicSubscriber): Unsubscribable {
+        this.subscribers.push(subscriber);
+
+        return {
+            unsubscribe: () => this.unsubscribe(subscriber)
+        };
+    }
+
+    unsubscribe(subscriber: TopicSubscriber): void {
+        const idx = this.subscribers.lastIndexOf(subscriber);
+
+        if (idx < 0)
+            return;
+
+        this.subscribers.splice(idx, 1);
+
+        if (this.subscribers.length == 0) {
+            this.onUnsubscribe();
+        }
+    }
+
+    next(message: TopicMessage): void {
+        for (let subscriber of this.subscribers) {
+            try {
+                subscriber.next?.call(subscriber, message);
+            }
+            catch (err) {
+                console.error(err);
```

Comment:
I don't know what's the preferred way for logging errors - is console.error() a safe default choice?

Reply:
Yes it is. If you log with the appropriate intent you can filter by log levels on the developer console, etc.